rfcs/0052-dynamic-ids.md

```diff
@@ -0,0 +1,121 @@
+- If applicable use systemd's `DynamicUser`
+- Otherwise let NixOS assign dynamic persistent ids, which happens automatically when `users.users.<name?>.uid`/`users.groups.<name?>.gid` is not set. For users, `users.users.<name?>.isSystemUser` should be set so that only uids under 1000 are used by NixOS services.
+
+Usage of static ids has to be explicitly justified.
+
+Note that this RFC is only about NixOS services in nixpkgs. It doesn't impose any restriction on the end users setting static ids. However note that without a central static uid mapping conflicts can occur, which will result in an error during the system build.
+
+# Motivation
+[motivation]: #motivation
+
+We are running out of static ids in the reserved range from 0 to 400. If services continue to reserve ids for themselves this will run out eventually.
+
+In addition, a central list of ids is annoying to maintain and leads to merge conflicts.
+
+# Detailed design
+[design]: #detailed-design
+
+## Documentation updates
+
+New best practices for declaring users are to be documented. This includes sections like the following
+
+### Using `DynamicUser` for declaring users
+
+If the service is fit for `DynamicUser`, this is the preferred solution, as it doesn't require any persistent ids. However this only works well for services whose state is self-contained. As soon as the service's data is needed somewhere else, this approach can fall flat. `DynamicUser` also enables a bunch of options restricting the service (see `man systemd.exec` for details), some of which might need to be disabled in order for it to work. Note that `DynamicUser` works even with large amounts of files, since it almost never has to change the underlying uid which would require a `chown` of all files.
+
+An example:
+
+```nix
+{
+  systemd.services.myservice.serviceConfig.DynamicUser = true;
+}
+```
+
+### Using ids dynamically allocated by NixOS
+
+
+By not setting `users.users.<name?>.uid`/`users.groups.<name?>.gid`, NixOS will dynamically allocate ids. For users, `users.users.<name?>.isSystemUser = true` should be set as well such that only ids below 1000 are used. These ids are persistent over the lifetime of a NixOS system, even when services are disabled and enabled again. The generated mapping from names to ids is stored in `/var/lib/nixos/uid-map`/`/var/lib/nixos/gid-map`, so if this directory is backed up, the mappings will persist too when restoring.
+
+An example:
+
+```nix
+{
+  users = {
+    users.myservice = {
+      description = "My service user";
+      group = "myservice";
+      isSystemUser = true;
+    };
+    groups.myservice = {};
+  };
+}
+```
+
+### Ensuring correct directory permissions for services
+
+Services must be able to access their directories.
+- The easiest way to achieve this is to use `systemd.services.<name?>.serviceConfig.StateDirectory = "myservice"`, which ensures that `/var/lib/myservice` belongs to the services user. See `man systemd.exec` for other related directives.
```

Comment:
Do we want to add some advice about avoiding including new `dataDir` options in newly created modules unless there's a very good usecase for it? (E.g. postgres)

Reply:
Out of scope for this RFC. I strongly agree with that opinion, but there are many who strongly disagree so we shouldn't muddy the waters here with that.

Reply:
Followup question.  Should we mention  `ConfigurationDirectory, CacheDirectory, LogsDirectory` here too? They all have the same kind of behaviour when `DynamicUser=true`  See https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectory=

Reply:
It's just a suggestion right? It doesn't say you *must* use `StateDirectory`.

Reply:
I definitely think we should be using `StateDirectory`, I just figured that might become a point of contention for some people. If everyone agrees then I guess there isn't a problem...

Reply:
I added a reference to the other directives in this sentence in https://github.com/NixOS/rfcs/pull/52/commits/ff7636a43fbb69da1b75f314b94e1d88634d7bc1. I feel like they deserve at least one mention.